gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/visitor/ExecutionVisitor.java

```diff
@@ -329,9 +311,81 @@ public EntityFetcherResponse visit(NoOpNode noOpNode) {
-  boolean isSingleSourceAndSame(Set<String> firstSource, Set<String> secondSource) {
-    return firstSource.size() == 1 &&
-        firstSource.equals(secondSource);
+  @Override
+  public EntityFetcherResponse visit(SelectionAndFilterNode selectionAndFilterNode) {
+    List<EntityFetcherResponse> resultMapList = new ArrayList<>();
+    EntitiesRequest request =
+        EntitiesRequest.newBuilder(executionContext.getEntitiesRequest())
+            .setOffset(selectionAndFilterNode.getOffset())
+            .setLimit(selectionAndFilterNode.getLimit())
+            .build();
+    EntitiesRequestContext context =
+        new EntitiesRequestContext(
+            executionContext.getTenantId(),
+            request.getStartTimeMillis(),
+            request.getEndTimeMillis(),
+            request.getEntityType(),
+            executionContext.getRequestHeaders());
+
+    String source = selectionAndFilterNode.getSource();
+    if (source.equals("QS")) {
+      // TODO: Make these queries parallel
+      IEntityFetcher entityFetcher = queryHandlerRegistry.getEntityFetcher(source);
+      // Entities Attributes and aggregations request
+      resultMapList.add(entityFetcher.getEntitiesAndAggregatedMetrics(context, request));
+      // Time Series request
+      resultMapList.add(entityFetcher.getTimeAggregatedMetrics(context, request));
+      return resultMapList.stream().reduce(new EntityFetcherResponse(), (r1, r2) -> union(Arrays.asList(r1, r2)));
+    } else { // EDS - can only get Entities, not Entity aggregations or time aggregations
+      IEntityFetcher entityFetcher = queryHandlerRegistry.getEntityFetcher(source);
+      return entityFetcher.getEntities(context, request);
+    }
+  }
+
+  @Override
+  public EntityFetcherResponse visit(PaginateOnlyNode paginateOnlyNode) {
+    EntityFetcherResponse result = paginateOnlyNode.getChildNode().acceptVisitor(this);
+
+    // Create a list from elements of HashMap
+    List<Map.Entry<EntityKey, Builder>> list =
+        new LinkedList<>(result.getEntityKeyBuilderMap().entrySet());
+
+    // Sort the list
+    List<Map.Entry<EntityKey, Entity.Builder>> sortedList =
+        DataCollectionUtil.paginateAndLimit(
+            list.stream(),
+            paginateOnlyNode.getLimit(),
+            paginateOnlyNode.getOffset());
+
+    // put data from sorted list to a linked hashmap
+    Map<EntityKey, Builder> linkedHashMap = new LinkedHashMap<>();
+    sortedList.forEach(entry -> linkedHashMap.put(entry.getKey(), entry.getValue()));
+    return new EntityFetcherResponse(linkedHashMap);
+  }
+
+  @Override
+  public EntityFetcherResponse visit(TotalFetcherNode totalFetcherNode) {
+    Future<EntityFetcherResponse> resultFuture = executorService.submit(() -> totalFetcherNode.getChildNode().acceptVisitor(this));
+    IEntityFetcher totalEntitiesFetcher = queryHandlerRegistry.getEntityFetcher(totalFetcherNode.getSource());
+
+    Future<Integer> totalFuture = executorService.submit(()->
```

Comment:
why can't total be combined into the original query as `COUNT(*)`?

Reply:
Because the EntitiesRequest query that goes into getEntitiesAndAggregatedMetrics() or getTimeAggregations() groups by entity ids. Adding a Count(*) does not give you what you want.